Pendulum/Data/CDPenPal+Extensions.swift

```diff
@@ -0,0 +1,270 @@
+//
+//  CDPenPal+Extensions.swift
+//  Pendulum
+//
+//  Created by Ben Cardy on 19/11/2022.
+//
+
+import Foundation
+import SwiftUI
+import CoreData
+import Contacts
+
+extension CDPenPal {
+    
+    static let entityName: String = "CDPenPal"
+    
+    var wrappedName: String {
+        self.name ?? "Unknown Pen Pal"
+    }
+    var wrappedInitials: String {
+        self.initials ?? "UP"
+    }
+ 
+    var lastEventType: EventType {
+        get { return EventType.from(self.lastEventTypeValue) }
+        set { self.lastEventTypeValue = Int16(newValue.rawValue) }
+    }
+    
+    var groupingEventType: EventType {
+        self.archived ? .archived : self.lastEventType
+    }
+    
+    var displayImage: Image? {
+        if let imageData = self.image, let image = UIImage(data: imageData) {
+            return Image(uiImage: image).resizable()
+        }
+        return nil
+    }
+    
+}
+
+extension CDPenPal {
+    
+    var ownEventsPredicate: NSPredicate {
+        NSPredicate(format: "penpal = %@", self)
+    }
+    
+    func addEvent(ofType eventType: EventType, date: Date? = Date(), notes: String? = nil, pen: String? = nil, ink: String? = nil, paper: String? = nil) {
+        dataLogger.debug("Adding event of type \(eventType.rawValue) to \(self.wrappedName)")
```

Comment:
Pending: Do we want to add logs to the persticenceController too? (Haven't reached that bit yet.

Reply:
The controller itself does very little, but more logging is always better than no logging!